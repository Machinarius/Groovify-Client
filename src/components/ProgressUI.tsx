import * as React from "react";

export interface IProps<TPromise> {
    promiseFactory: () => Promise<TPromise>;
    callback: (result: TPromise, error: Error) => void;
}

interface IState {
    promiseIsRunning: boolean,
    promiseRejected: boolean,
    promiseResolved: boolean
}

// TODO: Integrate https://github.com/lifeomic/attempt to provide resiliency/retry services
export default class ProgressUI<TPromise> extends React.Component<IProps<TPromise>, IState> {
    constructor(props: IProps<TPromise>) {
        super(props);

        this.startResolvingPromise = this.startResolvingPromise.bind(this);

        this.state = {
            promiseIsRunning: false,
            promiseRejected: false,
            promiseResolved: false
        };
    }

    componentDidMount() {
        this.startResolvingPromise();
    }

    private async startResolvingPromise(): Promise<any> {
        this.setState({
            promiseIsRunning: true
        });

        try {
            let result = await this.props.promiseFactory();
            this.props.callback(result, null);

            this.setState({
                promiseIsRunning: false,
                promiseResolved: true
            });
        } catch (error) {
            this.props.callback(null, error);

            this.setState({
                promiseIsRunning: false,
                promiseRejected: true
            });
        }
    }

    render() {
        if (this.state.promiseIsRunning) {
            return (
                <div>
                    Loading data...
                    <progress></progress>
                </div>
            );
        }

        if (this.state.promiseRejected) {
            return (
                <div>
                    <p>Loading data failed, please check your internet connection and try again</p>
                    <button className={"retry-button"} onClick={() => this.startResolvingPromise()}>Try Again</button>
                </div>
            );
        }

        let hiddenStyle = {
            display: "none"
        };

        return (<div style={hiddenStyle}></div>);
    }
}