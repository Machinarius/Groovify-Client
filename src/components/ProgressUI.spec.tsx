/**
 * @jest-environment jsdom
 */

import { mock, verify } from "ts-mockito";
import JestMockPromise from "jest-mock-promise";

import * as React from "react";
import * as testRenderer from "react-test-renderer";
import { shallow } from "enzyme";

import ProgressUI from "./ProgressUI";

test("Must show a progress bar element while the promise is resolving and then hide its own UI when the promise resolves", async () => {
    let mockPromise = new JestMockPromise();
    let fakePromise = mockPromise as any as Promise<number>;
    let fakePromiseFn = () => fakePromise;
    let component = testRenderer.create(
        <ProgressUI promiseFactory={fakePromiseFn} callback={(_result, _error) => {}} />
    );

    expect(component.toJSON()).toMatchSnapshot();
    
    mockPromise.resolve(42);
    await fakePromise;
    
    expect(component.toJSON()).toMatchSnapshot();
});

test("Must show a progress bar element while the promise is resolving and then show an error message when the promise rejects", async () => {
    let mockPromise = new JestMockPromise();
    let fakePromise = mockPromise as any as Promise<number>;
    let fakePromiseFn = () => fakePromise;
    let component = testRenderer.create(
        <ProgressUI promiseFactory={fakePromiseFn} callback={(_result, _error) => {}} />
    );

    expect(component.toJSON()).toMatchSnapshot();
    
    let expectedError = new Error("This is expected");
    mockPromise.reject(expectedError);
    try { await mockPromise; } catch (e) { }
    
    expect(component.toJSON()).toMatchSnapshot();
});

test("Must call the promise factory function and deliver the result into the callback", async () => {
    var deliveredResult: number;

    let mockPromise = new JestMockPromise();
    let fakePromise = mockPromise as any as Promise<number>;
    let fakePromiseFn = () => fakePromise;
    let component = testRenderer.create(
        <ProgressUI promiseFactory={fakePromiseFn} callback={(result, _error) => { deliveredResult = result; }} />
    );

    let expectedResult = 42;
    mockPromise.resolve(expectedResult);
    await fakePromise;

    expect(deliveredResult).toEqual(expectedResult);
});

test("Must call the promise factory function and deliver the error into the callback", async () => {
    var deliveredError: Error;

    let mockPromise = new JestMockPromise();
    let fakePromise = mockPromise as any as Promise<number>;
    let fakePromiseFn = () => fakePromise;
    let component = testRenderer.create(
        <ProgressUI promiseFactory={fakePromiseFn} callback={(_result, error) => { deliveredError = error; }} />
    );

    let expectedError = new Error("This is expected");
    mockPromise.reject(expectedError);
    try { await mockPromise; } catch (e) { }

    expect(deliveredError).toBe(expectedError);
});

test("Must retry calling the promise when the user clicks retry", async () => {
    var callCount = 0;

    let mockPromises = [new JestMockPromise(), new JestMockPromise()];
    let fakePromiseFn = () => {
        return mockPromises[callCount++] as any as Promise<number>;
    };
    let component = shallow(
        <ProgressUI promiseFactory={fakePromiseFn} callback={(_result, _error) => { }} />
    );

    let expectedError = new Error("This is expected");
    mockPromises[0].reject(expectedError);
    try { await mockPromises[0]; } catch (e) { }
    
    component.find('button.retry-button').simulate('click');
    expect(callCount).toEqual(2);
});