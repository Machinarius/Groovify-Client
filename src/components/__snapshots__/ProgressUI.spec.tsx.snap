// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Must show a progress bar element while the promise is resolving and then hide its own UI when the promise resolves 1`] = `
<div>
  Loading data...
  <progress />
</div>
`;

exports[`Must show a progress bar element while the promise is resolving and then hide its own UI when the promise resolves 2`] = `
<div
  style={
    Object {
      "display": "none",
    }
  }
/>
`;

exports[`Must show a progress bar element while the promise is resolving and then show an error message when the promise rejects 1`] = `
<div>
  Loading data...
  <progress />
</div>
`;

exports[`Must show a progress bar element while the promise is resolving and then show an error message when the promise rejects 2`] = `
<div>
  <p>
    Loading data failed, please check your internet connection and try again
  </p>
  <button
    className="retry-button"
    onClick={[Function]}
  >
    Try Again
  </button>
</div>
`;
